lib/Sema/CodeSynthesisDistributedActor.cpp

```diff
@@ -81,9 +96,519 @@ static VarDecl *addImplicitDistributedActorIDProperty(
+
+  DeclName thunkName = createDistributedFuncName(C, func);
+
+  // --- Prepare generic parameters
+  GenericParamList *genericParamList = nullptr;
+  if (auto genericParams = func->getGenericParams()) {
+    genericParamList = genericParams->clone(DC);
+  }
+
+  GenericSignature thunkGenSig =
+      buildGenericSignature(C, func->getGenericSignature(),
+                            /*addedParameters=*/{},
+                            /*addedRequirements=*/{});
+
+  // --- Prepare parameters
+  auto funcParams = func->getParameters();
+  SmallVector<ParamDecl*, 2> paramDecls;
+  for (unsigned i : indices(*func->getParameters())) {
+    auto funcParam = funcParams->get(i);
+
+    auto paramName = funcParam->getParameterName();
+    // If internal name is empty it could only mean either
+    // `_:` or `x _: ...`, so let's auto-generate a name
+    // to be used in the body of a thunk.
+    if (paramName.empty()) {
+      paramName = C.getIdentifier("p" + llvm::utostr(i));
+    }
+
+    auto paramDecl = new (C)
+        ParamDecl(SourceLoc(),
+                  /*argumentNameLoc=*/SourceLoc(), funcParam->getArgumentName(),
+                  /*parameterNameLoc=*/SourceLoc(), paramName, DC);
+
+    paramDecl->setImplicit(true);
+    paramDecl->setSpecifier(funcParam->getSpecifier());
+    paramDecl->setInterfaceType(funcParam->getInterfaceType());
+
+    paramDecls.push_back(paramDecl);
+  }
+  ParameterList *params = ParameterList::create(C, paramDecls); // = funcParams->clone(C);
+
+    auto thunk = FuncDecl::createImplicit(C, swift::StaticSpellingKind::None,
+                           thunkName, SourceLoc(),
+                           /*async=*/true, /*throws=*/true,
+                           genericParamList,
+                           params,
+                           func->getResultInterfaceType(),
+                           DC);
+  thunk->setSynthesized(true);
+  thunk->getAttrs().add(new (C) NonisolatedAttr(/*implicit=*/true));
+  thunk->setGenericSignature(thunkGenSig);
+  thunk->copyFormalAccessFrom(func, /*sourceIsParentContext=*/false);
+  thunk->setBodySynthesizer(deriveBodyDistributed_thunk, func);
+
+  return thunk;
+}
+
+/******************************************************************************/
+/*********************** SYNTHESIS ENTRY POINTS *******************************/
+/******************************************************************************/
+
+FuncDecl *GetDistributedThunkRequest::evaluate(
+    Evaluator &evaluator, AbstractFunctionDecl *afd) const {
+  if (!afd->isDistributed())
+    return nullptr;
+
+  auto &C = afd->getASTContext();
+  auto DC = afd->getDeclContext();
+
+  if (!getConcreteReplacementForProtocolActorSystemType(afd)) {
+    // Don't synthesize thunks, unless there is a *concrete* ActorSystem.
+    // TODO(distributed): we should be able to lift this eventually,
+    // and allow resolving distributed actor protocols.
```

Comment:
This has to emit an error message.

Reply:
Hmm, we issue this request also for `distributed func` inside protocols and then we just return nothing I think.

Erroring here would cause us to reject

```
protocol NoSerializationRequirementYet: DistributedActor {
  // OK, no serialization requirement yet
  distributed func test() -> NotCodable
}
```

i think this protocol is fine 🤔 

We should reject such extension though I guess... That sounds right, I'll do that separately

Reply:
It sounds like the `if` here is actually an assertion and caller has to be detect and diagnose that something doesn't have a concrete actor system?

Reply:
Yeah I think you're right, we should not call this in protocols I guess.

Will follow up with a patch for that. Then we can diagnose here and prevent bad extensions 👍

Reply:
Sounds good!